src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1559 @@
+    SuccessOrExit(error = Dns::Name::AppendTerminator(aMessage));
+
+    offset = aMessage.GetLength();
+    SuccessOrExit(error = aMessage.Append(sig));
+    SuccessOrExit(error = AppendHostName(aMessage, aInfo));
+    SuccessOrExit(error = aMessage.Append(signature));
+    UpdateRecordLengthInMessage(sig, offset, aMessage);
+
+exit:
+    return error;
+}
+
+void Client::UpdateRecordLengthInMessage(Dns::ResourceRecord &aRecord, uint16_t aOffset, Message &aMessage) const
+{
+    // This method is used to calculate an RR DATA length and update
+    // (rewrite) it in a message. This should be called immediately
+    // after all the fields in the record are written in the message.
+    // `aOffset` gives the offset in the message to the start of the
+    // record.
+
+    aRecord.SetLength(aMessage.GetLength() - aOffset - sizeof(Dns::ResourceRecord));
+    aMessage.Write(aOffset, aRecord);
+}
+
+void Client::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    static_cast<Client *>(aContext)->ProcessResponse(*static_cast<Message *>(aMessage));
+}
+
+void Client::ProcessResponse(Message &aMessage)
+{
+    static const ItemState kNewStateOnUpdateDone[]{
+        /* (0) kToAdd      -> */ kToAdd,
+        /* (1) kAdding     -> */ kRegistered,
+        /* (2) kToRefresh  -> */ kToRefresh,
+        /* (3) kRefreshing -> */ kRegistered,
+        /* (4) kToRemove   -> */ kToRemove,
+        /* (5) kRemoving   -> */ kRemoved,
+        /* (6) kRegistered -> */ kRegistered,
+        /* (7) kRemoved    -> */ kRemoved,
+    };
+
+    otError             error = OT_ERROR_NONE;
+    Dns::UpdateHeader   header;
+    uint16_t            offset = aMessage.GetOffset();
+    uint16_t            recordCount;
+    LinkedList<Service> removedServices;
+
+    VerifyOrExit(GetState() == kStateUpdating);
+
+    SuccessOrExit(error = aMessage.Read(offset, header));
+
+    VerifyOrExit(header.GetType() == Dns::Header::kTypeResponse, error = OT_ERROR_PARSE);
+    VerifyOrExit(header.GetQueryType() == Dns::Header::kQueryTypeUpdate, error = OT_ERROR_PARSE);
+    VerifyOrExit(header.GetMessageId() == mUpdateMessageId, error = OT_ERROR_DROP);
+
+    if (!Get<Mle::Mle>().IsRxOnWhenIdle())
+    {
+        Get<DataPollSender>().StopFastPolls();
+    }
+
+    // Response is for the earlier request message.
+
+    otLogInfoSrp("[client] Received response");
+
+    error = Dns::Header::ResponseCodeToError(header.GetResponseCode());
+
+    if (error != OT_ERROR_NONE)
+    {
+        otLogInfoSrp("[client] Server rejected %s code:%d", otThreadErrorToString(error), header.GetResponseCode());
+
+        if (mHostInfo.GetState() == kAdding)
+        {
+            // Since server rejected the update message, we go back to
+            // `kToAdd` state to allow user to give a new name using
+            // `SetHostName()`.
+            mHostInfo.SetState(kToAdd);
```

Comment:
I think the caller needs to provide another name only when there is a name conflict (i.e. response code is `kResponseNameExists`).

Another issue with name conflicts is that there is no way for the SRP server to specify which name has been already taken. So the client cannot correctly notify the caller whether a new host or service instance name is needed. I think one solution is to have the SRP server includes the conflicting `Instruction` (or just a single "Delete All" RR which includes the conflicting name) in the response, so that the client knows which name it needs to update.
Thoughts? @abtink @Abhayakara @jwhui 

FYI, latest description about name conflicting: https://tools.ietf.org/html/draft-ietf-dnssd-srp-09#section-2.2.5.2

Reply:
My assumption has been that you wouldn’t send multiple services at once, but I think we’ve blown past that restriction. So yes, sending back the instructions that failed could work.  Let’s discuss on the wg mailing list?

Reply:
Yes, spec discussion should occur on the IETF WG mailer.

Reply:
> My assumption has been that you wouldn’t send multiple services at once, but I think we’ve blown past that restriction. So yes, sending back the instructions that failed could work. Let’s discuss on the wg mailing list?

Sure! Just came up with this issue when reading the implementation, will post on the mailing list.